linker/shared/src/main/scala/org/scalajs/linker/standard/LinkTimeProperties.scala

```diff
@@ -0,0 +1,74 @@
+package org.scalajs.linker.standard
+
+import org.scalajs.ir.Trees.LinkTimeTree
+import org.scalajs.ir.Trees.LinkTimeOp
+import org.scalajs.linker.interface.{Semantics, ESFeatures}
+
+final class LinkTimeProperties (
+  semantics: Semantics,
+  esFeatures: ESFeatures
+) {
+  import LinkTimeProperties._
+  import LinkTimeProperties.ResolvedLinkTimeTree._
+
+  private val linkTimeProperties: Map[String, ResolvedLinkTimeTree] = Map(
+    // Must be in sync with the arguments of @linkTimeProperty("...")
+    // for the fields in `scala.scalajs.LinkingInfo`.
+    "core/productionMode" -> BooleanValue(semantics.productionMode),
+    "core/esVersion"      -> IntValue(esFeatures.esVersion.edition)
+  )
+
+  def evaluateLinkTimeTree(cond: LinkTimeTree): Boolean = {
+    eval(cond) match {
+      case BooleanValue(v) => v
+      case IntValue(v) =>
+        throw new IllegalArgumentException(
+            "Link-time condition must be evaluated to be a boolean value, but int is found.")
+    }
+  }
+
+  private def eval(cond: LinkTimeTree): ResolvedLinkTimeTree = cond match {
+    case LinkTimeTree.BinaryOp(op, lhs, rhs) =>
+      ResolvedLinkTimeTree.BooleanValue {
+        (eval(lhs), eval(rhs)) match {
+          case (IntValue(l), IntValue(r)) =>
+            op match {
+              case LinkTimeOp.Int_== => l == r
+              case LinkTimeOp.Int_!= => l != r
+              case LinkTimeOp.Int_<  => l < r
+              case LinkTimeOp.Int_<= => l <= r
+              case LinkTimeOp.Int_>  => l > r
+              case LinkTimeOp.Int_>= => l >= r
+              case _ =>
+                throw new IllegalArgumentException(s"Invalid operation $op for int values.")
+            }
+          case (BooleanValue(l), BooleanValue(r)) =>
+            op match {
+              case LinkTimeOp.Boolean_== => l == r
+              case LinkTimeOp.Boolean_!= => l != r
+              case LinkTimeOp.Boolean_|| => l || r
+              case LinkTimeOp.Boolean_&& => l && r
+              case _ =>
+                throw new IllegalArgumentException(s"Invalid operation $op for boolean values.")
+            }
+          case _ =>
+            throw new IllegalArgumentException("Type mismatch: binary operation with different types " +
+              "is not allowed in linkTimeIf.")
+        }
+      }
+    case LinkTimeTree.BooleanConst(v)   => BooleanValue(v)
+    case LinkTimeTree.IntConst(v)       => IntValue(v)
+    case LinkTimeTree.Property(name, _) => resolveLinkTimeProperty(name)
+  }
+
+  private def resolveLinkTimeProperty(prop: String): ResolvedLinkTimeTree =
+    linkTimeProperties.getOrElse(prop, throw new IllegalArgumentException(s"link time property not found: '$prop'"))
```

Comment:
I'm currently handling invalid `LinkTimeTree` input by throwing an `IllegalArgumentException` here and there.
These exceptions shouldn't be user-facing, and they indicates either a compiler/linker error or that we've forgotten to add a pre-defined `linkTimeProperties` as long as we don't allow users to define custom link-time properties.
Invalid input should be validated and kept out in the earlier stages.

I'm wondering throwing an exception is a right way to handle invalid inputs here? Are there better way to expressing the link-time errors than throwing an exceptions?

Reply:
The proper way to report invalid inputs is through ClassDefChecker errors or IRChecker errors. IR that should never be produced by a correct compiler for a correct codebase, but which is serializable/representable in the data structures, should get errors reported through either one of those. So for example, if currently we say there is a fixed set of properties that are valid, then `ClassDefChecker` should report diagnostics for `LinkTimeTree.Property` that mention unknown properties.

An `IllegalArgumentException` is fine for an incorrect *config* passed to the linker.

Reply:
>  ClassDefChecker should report diagnostics for LinkTimeTree.Property that mention unknown properties.

Thanks, right :+1:
`LinkTimeProperty` resolution checks in `ClassDefChecker` may remain valuable even after we allow users to define user-defined link-time conditions, because `ClassDefChecker` should know those custom link-time properties through user configuration anyway.